C9 Support for understanding-change-sets (#218)
aws-samples/cfn101-workshop: content/intermediate/operations/understanding-changesets/index.md

Reviewer: There may be more `changesets-workshop` that I couldn't suggest changes here

Let’s get started!

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}

>     1. Change directory to: `code/workspace/understanding-changesets`.

Author: the working directory for Cloud9 environment is bit different. I have updated the new directory

        1. Let's create the stack from the template using the following command (the example uses `us-east-1` for the AWS region, change this value as needed):
        :::code{language=shell showLineNumbers=false showCopyAction=true}
        aws cloudformation create-stack \
            --region us-east-1 \
            --stack-name changesets-workshop \
            --template-body file://bucket.yaml \
            --parameters ParameterKey=BucketName,ParameterValue='TYPE_UNIQUE_BUCKET_NAME-HERE'
        :::
    1. CloudFormation returns the following output:
    :::code{language=shell showLineNumbers=false showCopyAction=true}
  {
  "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/changesets-workshop/330b0120-1771-11e4-af37-50ba1b98bea6"
}
    :::
    1. Wait until the `changesets-workshop` stack is created, by using the CloudFormation console or the [stack-create-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-create-complete.html) wait command of the AWS CLI 
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation wait stack-create-complete \

>       --stack-name changesets-workshop \

Reviewer: ```suggestion
      --stack-name changesets-workshop
```

Let’s get started!

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}
	
    1. Change directory to: `cfn101-workshop/code/workspace/understanding-changesets`.

>     1. Open the `bucket.yaml` CloudFormation template in your `Cloud9` editor, and familiarize yourself with the sample template content.

Reviewer: ```suggestion
    1. Open the `bucket.yaml` CloudFormation template in your **Cloud9 editor**, and familiarize yourself with the sample template content.
```

    1. Create a stack by following these steps:
    
        1. The template requires you to provide a unique value for the `BucketName` input parameter. For more information, see [Bucket naming rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html)
        1. Let's create the stack from the template using the following command (the example uses `us-east-1` for the AWS region, change this value as needed):
        :::code{language=shell showLineNumbers=false showCopyAction=true}
        aws cloudformation create-stack \
                --region us-east-1 \

>                 --stack-name changesets-workshop \

Reviewer: ```suggestion
                --stack-name cfn-workshop-understanding-changesets \
```

                --template-body file://bucket.yaml \
                --parameters ParameterKey=BucketName,ParameterValue='YOUR_UNIQUE_BUCKET_NAME-HERE'
        :::
        1. CloudFormation returns the following output:

>             ```json

Reviewer: ```suggestion
            :::code{language=json showLineNumbers=true showCopyAction=false}
```

            {

>             "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/changesets-workshop/330b0120-1771-11e4-af37-50ba1b98bea6"

Reviewer: ```suggestion
            "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-understanding-changesets/330b0120-1771-11e4-af37-50ba1b98bea6"
```

            }            

>             ```

Reviewer: ```suggestion
            :::
```

>         1. Wait until the `changesets-workshop` stack is created, by using the CloudFormation console or the [wait stack-create-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-create-complete.html) AWS CLI command.

Reviewer: ```suggestion
        1. Wait until the `cfn-workshop-changesets` stack is created, by using the CloudFormation console or the [wait stack-create-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-create-complete.html) AWS CLI command.
```

>         1. Wait until the `changesets-workshop` stack is created, by using the CloudFormation console or the [wait stack-create-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-create-complete.html) AWS CLI command.

Reviewer: ```suggestion
        1. Wait until the `cfn-workshop-understanding-changesets` stack is created, by using the CloudFormation console or the [wait stack-create-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-create-complete.html) AWS CLI command.
```

        :::code{language=shell showLineNumbers=false showCopyAction=true}
        aws cloudformation wait stack-create-complete \

>     	        --stack-name changesets-workshop

Reviewer: ```suggestion
    	        --stack-name cfn-workshop-understanding-changesets
```

    8. In the review page, choose **Create Stack**.
    9. Refresh the stack creation page until you see your stack in the `CREATE_COMPLETE` status.

   ::::
   :::::
### Lab part 1
In this part of the lab, you will specify a property, for a given resource type, that requires [no interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt) on stack updates. You will then create a change set to preview the changes, and inspect the output of the change set operation.

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}

> Open the `bucket.yaml` CloudFormation template in your `Cloud9` editor, and add `VersioningConfiguration` as shown below. Save the file.

Reviewer: Suggest moving the open/edit of template outside the tabbed instructions

Author: the instructions were specific to Cloud9 and it is different for local development.

```yaml
MyS3Bucket:
  Type: AWS::S3::Bucket
  Properties:
    BucketName: !Ref BucketName
    VersioningConfiguration:
      Status: Enabled
```
Next, create your first change set:

    1. From Terminal run the following command to **Create change set**, specify a name for the change set for example:`bucket-versioning-update` and provide the `BucketName` parameter that chosen in above example.
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation create-change-set \
        --stack-name changesets-workshop \

>         --change-set-name bucket-versioning-update

Reviewer: ```suggestion
        --change-set-name bucket-versioning-update \
```

        --template-body file://bucket.yaml \
        --parameters ParameterKey=BucketName,ParameterValue='TYPE_UNIQUE_BUCKET_NAME-HERE'
    :::
    1. CloudFormation returns the following output of the AWS CLI
    :::code{language=shell showLineNumbers=false showCopyAction=true}
  {

>   "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/changesets-workshop/330b0120-1771-11e4-af37-50ba1b98bea6",

Reviewer: these lines are swapped in my testing

```suggestion
    "Id": "arn:aws:cloudformation:us-east-1:123456789012:changeSet/bucket-versioning-update/a470cff7-cb2c-4cba-bf27-2b3b9ccc1333",
```

>   "Id": "arn:aws:cloudformation:us-east-1:123456789012:changeSet/bucket-versioning-update/a470cff7-cb2c-4cba-bf27-2b3b9ccc1333"

Reviewer: ```suggestion
  "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/changesets-workshop/330b0120-1771-11e4-af37-50ba1b98bea6"
```

} 
   :::

>     1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).

Reviewer: Also suggest moving this entire "navigate to the console and look at the changeset JSON" to outside the tabbed instructions - it's the same either way

```suggestion
    1. In the [AWS CloudFormation ](https://console.aws.amazon.com/cloudformation/) console, select the stack you created in this workshop. For example `changesets-workshop`.
```

>     1. From **stacks**, choose **changesets-workshop**.

Reviewer: ```suggestion
```

>     1. From **changesets**, choose **bucket-versioning-update**. 

Reviewer: ```suggestion
    1. From **changesets** tab, choose **bucket-versioning-update**. 
```

>     8. In the review page, choose **Create Stack**.

Reviewer: ```suggestion
    8. In the review page, choose **Submit**.
```

   ::::
   :::::

### Lab part 1
In this part of the lab, you will specify a property, for a given resource type, that requires [no interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt) on stack updates. You will then create a change set to preview the changes, and inspect the output of the change set operation.

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}
	
Open the `bucket.yaml` CloudFormation template in your `Cloud9` editor, and add `VersioningConfiguration` as shown below. Save the file.

```yaml
MyS3Bucket:
  Type: AWS::S3::Bucket
  Properties:
    BucketName: !Ref BucketName
    VersioningConfiguration:
      Status: Enabled
```
Next, create your first change set:

    1. From Terminal run the following command to **Create change set**, specify a name for the change set for example:`bucket-versioning-update` and provide the `BucketName` parameter that chosen in above example.
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    
    aws cloudformation create-change-set \

>         --stack-name changesets-workshop \

Reviewer: ```suggestion
        --stack-name cfn-workshop-understanding-changesets \
```

        --change-set-name bucket-versioning-update
        --template-body file://bucket.yaml \
        --parameters ParameterKey=BucketName,ParameterValue='TYPE_UNIQUE_BUCKET_NAME-HERE'
    :::

    2. CloudFormation returns the following output of the AWS CLI.

>         ```json

Reviewer: ```suggestion
        :::code{language=json showLineNumbers=true showCopyAction=false}
```

        {

>         "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/changesets-workshop/330b0120-1771-11e4-af37-50ba1b98bea6",

Reviewer: ```suggestion
        "StackId" : "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-understanding-changesets/330b0120-1771-11e4-af37-50ba1b98bea6",
```

        "Id": "arn:aws:cloudformation:us-east-1:123456789012:changeSet/bucket-versioning-update/a470cff7-cb2c-4cba-bf27-2b3b9ccc1333"
        }

>         ```

Reviewer: ```suggestion
        :::
```

Let’s get started!

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}

1. From Terminal run the following command to **Create change set**, Change the value for `BucketName` parameter by specifying a new unique bucket [name](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), and follow the rest of the process as before to finish creating the change set.

 :::code{language=shell showLineNumbers=false showCopyAction=true}
 aws cloudformation create-change-set \

>       --stack-name changesets-workshop \

Reviewer: ```suggestion
      --stack-name cfn-workshop-understanding-changesets \
```

      --change-set-name replace-change-set \
      --template-body file://bucket.yaml \
      --parameters ParameterKey=BucketName,ParameterValue='YOUR-NEW-UNIQUE-BUCKET-NAME-HERE'
 :::

  ::::
	::::tab{id="local" label="Local development"}

>     1. In the CloudFormation console, select the `changesets-workshop` stack, and from **Stack actions**, choose **Create change set for current stack**.

Reviewer: ```suggestion
    1. In the CloudFormation console, select the `cfn-workshop-understanding-changesets` stack, and from **Stack actions**, choose **Create change set for current stack**.
```

Let’s get started!

   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}
  From Terminal run the following command to **Create change set**, Change the value for `BucketName` parameter by specifying a new unique bucket [name](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), and follow the rest of the process as before to finish creating the change set.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation create-change-set \
      --stack-name changesets-workshop \
      --change-set-name replace-change-set
      --template-body file://bucket.yaml \
      --parameters ParameterKey=BucketName,ParameterValue='NEW_UNIQUE_BUCKET_NAME-HERE'
  :::

>   

Author: This was left out purpose since the instructions were given to follow earlier process to review the JSON. I am open to add the output if required

>   Here’s what the **JSON changes** for this change set should look like:

Reviewer: suggest moving this entire "navigate to the console and look at the changeset JSON" to outside the tabbed instructions - it's the same either way

### Cleanup

To clean up resources you created with this lab:
   :::::tabs{variant="container"}
	::::tab{id="cloud9" label="Cloud9"}
	From the terminal execute the following AWS CLI command to delete the stacks you created.
 :::code{language=shell showLineNumbers=false showCopyAction=true}
 aws cloudformation delete-stack \

>       --stack-name changesets-workshop

Reviewer: ```suggestion
      --stack-name cfn-workshop-understanding-changesets
```

 :::
  ::::
	::::tab{id="local" label="Local development"}  

>   1. From the CloudFormation console, select the stack named `changesets-workshop`.

Reviewer: ```suggestion
  1. From the CloudFormation console, select the stack named `cfn-workshop-understanding-changesets`.
```
